qiita_pet/handlers/download.py

```diff
@@ -37,19 +38,42 @@ def get(self, filepath_id):
                 "filepath_id: %s" % (self.current_user.email, str(fid)))
 
         relpath = filepath_id_to_rel_path(fid)
+        fp_info = get_filepath_information(fid)
         fname = basename(relpath)
 
-        # If we don't have nginx, write a file that indicates this
-        self.write("This installation of Qiita was not equipped with nginx, "
-                   "so it is incapable of serving files. The file you "
-                   "attempted to download is located at %s" % relpath)
+        if fp_info['filepath_type'] in ('directory', 'html_summary_dir'):
+            # This is a directory, we need to list all the files so NGINX
+            # can download all of them
+            basedir = get_db_files_base_dir()
+            basedir_len = len(basedir) + 1
+            to_download = []
+            for dp, _, fps in walk(fp_info['fullpath']):
+                for fn in fps:
+                    fullpath = join(dp, fn)
+                    spath = fullpath
+                    if fullpath.startswith(basedir):
+                        spath = fullpath[basedir_len:]
+                    to_download.append((fullpath, spath, spath))
+
+            all_files = '\n'.join(
+                ["- %s /protected/%s %s" % (getsize(fp), sfp, n)
+                 for fp, sfp, n in to_download])
+
+            self.set_header('X-Archive-Files', 'zip')
+            self.write("%s\n" % all_files)
+            fname = '%s.zip' % fname
+        else:
+            # If we don't have nginx, write a file that indicates this
```

Comment:
A lot of this code is copy pastes from other classes, would you mind creating a new class from which these functions are inherit? Perhaps use the BaseHandlerDownload handler ...